### DISCUSSION TITLE
Trying to hit the Service Now API using basic auth

### DISCUSSION THREAD

**Discussion by LongBeachHXC** [encode/httpx#1553] [Potential Issue]
Hello All,

I'm having issues hitting the service now API using httpx basic auth. I've tested API connectivity with Postman and I can connect with no issues. What I notice is the Authorization header that I get from postman is dramatically different than the header I get with httpx or python in general. I'm a bit perplexed as to why this is as the basic auth appears pretty simple.

I must be missing something, some assistance pointing me in the right direction would be much appreciated.

Thank you.

**Comment by StephenBrown2** [comment]
Can you provide examples of the differences you're seeing? Of course with sensitive information redacted, but it'll be a lot easier to help with your example httpx code that doesn't work, and the HTTP session from Postman that does.

**Comment by LongBeachHXC** [comment]
Very simple request out to Service Now. I used Postman first to model out the request and ensure I have all my headers and params correct. Once I was able to get the data I wanted in postman I then went to python. The only real difference I can tell between the two is the base64 encoded authorization header. When I execute the _build_auth_header function in the python interpreter the token I get back is different than the token that is generated in postman. When I hardcode the authorization header with the token from postman everything works as expected.

```
>>> with httpx.Client(base_url=base_uri, auth=('username','password')) as client:
...     request = client.build_request('GET', cmdb_table, params=params, headers=headers)
...     resp = client.send(request)
... 

>>> resp
<Response [401 Unauthorized]>
```

**Reply by StephenBrown2** [reply]
Do you have MFA enabled? https://docs.servicenow.com/bundle/quebec-application-development/page/integrate/inbound-rest/concept/c_RESTAPI.html#d955432e652 The token would change between requests, I presume.
HTTPX does base64 the credentials, so I'd be surprised if it was significantly different if the inputs were exactly the same:
```pycon
In [1]: import base64

In [2]: import json

In [3]: import httpx

In [4]: with httpx.Client(base_url='https://httpbin.org', auth=('username', 'password')) as client:
   ...:     print("Client Auth:", client.auth.__dict__)
   ...:     r = client.get('/headers')
   ...:     print("Sent Headers:", json.dumps(r.json(), indent=2))
   ...:     print("Request Auth:", base64.b64decode(r.request.headers["Authorization"].removeprefix("Basic ")).decode())
   ...:     print("Response Auth:", base64.b64decode(r.json()["headers"]["Authorization"].removeprefix("Basic ")).decode())
   ...: 
Client Auth: {'_auth_header': 'Basic dXNlcm5hbWU6cGFzc3dvcmQ='}
Sent Headers: {
  "headers": {
    "Accept": "*/*",
    "Accept-Encoding": "gzip, deflate",
    "Authorization": "Basic dXNlcm5hbWU6cGFzc3dvcmQ=",
    "Host": "httpbin.org",
    "User-Agent": "python-httpx/0.17.1",
    "X-Amzn-Trace-Id": "Root=1-6075bf5a-19cae5cd47baf21c06714032"
  }
}
Request Auth: username:password
Response Auth: username:password
```

**Reply by LongBeachHXC** [reply]
No, MFA is not enabled. 

I double checked postman to see what auth type and creds I am using. My postman call is using basic auth and the same username/password combo that I am passing in to httpx.

I don't know if this problem is specific to httpx. It seems like python isn't encoding my username:password combo the same way that postman is. I am getting different values between the two.

**Reply by StephenBrown2** [reply]
Try plugging your Postman header into `base64.b64decode()` and see what it spits out?
